spyder/api/folding.py

```diff
@@ -0,0 +1,365 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © Spyder Project Contributors
+# Copyright © <2013-2016> <Colin Duquesnoy and others, see pyqode/AUTHORS.rst>
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+
+"""
+This module contains the code folding API.
+Adapted from https://github.com/pyQode/pyqode.core/blob/master/pyqode/core/api/folding.py
+"""
+from __future__ import print_function
+import sys
+from spyder.utils.editor import TextBlockHelper
+
+
+def print_tree(editor, file=sys.stdout, print_blocks=False):
+    """
+    Prints the editor fold tree to stdout, for debugging purpose.
+
+    :param editor: CodeEditor instance.
+    :param file: file handle where the tree will be printed. Default is stdout.
+    :param print_blocks: True to print all blocks, False to only print blocks
+        that are fold triggers
+    """
+    block = editor.document().firstBlock()
+    while block.isValid():
+        trigger = TextBlockHelper().is_fold_trigger(block)
+        trigger_state = TextBlockHelper().is_collapsed(block)
+        lvl = TextBlockHelper().get_fold_lvl(block)
+        visible = 'V' if block.isVisible() else 'I'
+        if trigger:
+            trigger = '+' if trigger_state else '-'
+            print('l%d:%s%s%s' %
+                  (block.blockNumber() + 1, lvl, trigger, visible),
+                  file=file)
+        elif print_blocks:
+            print('l%d:%s%s' %
+                  (block.blockNumber() + 1, lvl, visible), file=file)
+        block = block.next()
+
+
+class FoldDetector(object):
+    """
+    Base class for fold detectors.
+
+    A fold detector takes care of detecting the text blocks fold levels that
+    are used by the FoldingPanel to render the document outline.
+
+    To use a FoldDetector, simply set it on a syntax_highlighter::
+
+        editor.syntax_highlighter.fold_detector = my_fold_detector
```

Comment:
Please add to this docstring this

    Some implementations of fold detectors can be found in
    ``spyder/widgets/sourcecode/folding.py``.